linkerd/http/retry/src/peek_trailers.rs

```diff
@@ -17,31 +18,34 @@ use std::{
 /// If the first frame of the body stream was *not* a `TRAILERS` frame, this
 /// behaves identically to a normal body.
 #[pin_project(project = Projection)]
-pub struct PeekTrailersBody<B: Body = BoxBody> {
-    /// The inner [`Body`].
+pub enum PeekTrailersBody<B: Body = BoxBody> {
```

Comment:
a note for review: this part of the diff is best viewed in a side-by-side view, or by reading the file itself in tree. we've largely replaced how we model `PeekTrailersBody<B>`, so this diff does not present itself particularly well.

as noted elsewhere, we cannot pattern match on a `Frame<T>`. between that, and the multiple edge-cases we are bound to encounter depending on the kinds of frames yielded, and the readiness of the second frame, storing fields like `first_data: Option<Result<Frame<B::Data>, B::Error>>` ended up causing significant headaches in previous experiments.

Reply:
This isn't a hard blocker, but it would generally be preferable to do something like

```
pub struct PeekTrailersBody<B: Body = BoxBody>(Inner<B>);

enum Inner<B> { ... }
```

so that the enum variants/fields are private.

Reply:
great idea. this is done in a3096ef4